plugin/src/cli_utils/upgrade/mod.rs

```diff
@@ -0,0 +1,258 @@
+    };
+
+    delete_older_upgrade_events(release_name.as_str(), "upgrade", Api::namespaced(k8s_client.clone(), namespace)).await?;
+
+    create_upgrade_resources(args, release_name.as_str(), namespace, k8s_client).await?;
+
+
+
+    Ok(())
+}
+
+async fn joined_flatten<T>(handle: JoinHandle<Result<T>>) -> Result<T> {
+    match handle.await.map_error(|err| anyhow!(err)) {
+        Ok(Ok(result)) => Ok(result),
+        Ok(Err(err)) => Err(err),
+        Err(err) => Err(err),
+    }
+}
+
+pub async fn create_upgrade_resources(args: &UpgradeArgs, release_name: &str, ns: &str, k8s_client: Client) -> Result<()> {
+    let sa = upgrade_job_service_account(Some(ns.to_string()), upgrade_name_concat(release_name, "upgrade-service-account"));
+    let sa_client: Api<ServiceAccount> = Api::namespaced(k8s_client.clone(), ns);
+    let sa_create = spawn(idempotent_create_resource(&sa_client, &sa));
+
+    let cluster_role_binding = upgrade_job_cluster_role_binding(Some(ns.to_string()), release_name.to_string());
+    let cluster_role_binding_client: Api<ClusterRoleBinding> = Api::all(k8s_client.clone());
+    let cluster_role_binding_create = spawn(idempotent_create_resource(&cluster_role_binding_client, &cluster_role_binding));
+
+    let cluster_role = upgrade_job_cluster_role(Some(ns.to_string()), upgrade_name_concat(release_name, "upgrade-cluster-role"));
+    let cluster_role_client: Api<ClusterRole> = Api::all(k8s_client.clone());
+    let cluster_role_create = spawn(idempotent_create_resource(&cluster_role_client, &cluster_role));
+
+    let cm_data = config_map_data(args.set_file.as_slice())?;
+    let cm = upgrade_configmap(cm_data.0, ns, release_name.to_string());
+    let cm_client: Api<ConfigMap> = Api::namespaced(k8s_client.clone(), ns);
+    let cm_create = spawn(idempotent_create_resource(&cm_client, &cm));
+
+    let set_file = job_set_file_args(args.set_file.as_slice(), Some(cm_data.1))?;
+    let job = upgrade_job(
+        args,
+        ns,
+        release_name,
+        set_file.unwrap_or_default(),
+    )?;
+    let job_client: Api<Job> = Api::namespaced(k8s_client.clone(), ns);
+    let job_create = spawn(idempotent_create_resource(&job_client, &job));
+
+
+    try_join!(
+        joined_flatten(sa_create),
+        joined_flatten(cluster_role_binding_create),
+        joined_flatten(cluster_role_create),
+        joined_flatten(cm_create),
+        joined_flatten(job_create),
+    )?;
+
+    Ok(())
+}
+
+pub async fn idempotent_create_resource<K>(client: &Api<K>, resource: &K) -> Result<()>
+where
+    K: Resource + Clone + std::fmt::Debug,
+{
+    let pp = PostParams::default();
+    client
+        .create(&pp, resource)
+        .await
+        .map(|_| ())
+        .or_else(|err| match err {
+            kubeError::Api(response) if response.reason.eq("AlreadyExists") => {
+                println!("{kind} '{name}' already exists", kind = resource.kind(), name = resource.name_unchecked());
```

Comment:
Shouldn't we check the contents match as well?
If the resources are specifically for upgrade, another way could be deleting them before hand, if they exist.

Reply:
The names are unique across versions. So, we should be okay to assume that the resources is the correct one. WDYT?

Reply:
Then should be ok I guess

Reply:
Actually it may not be, ex: what if this content is from a previously failed upgrade?

Reply:
There are delete command to delete the resources. And it'll be printed that it already exists, which implies that creation was skipped?